exercises/src/main/scala/fpinscala/datastructures/List.scala

```diff
@@ -49,20 +51,224 @@ object List { // `List` companion object. Contains functions for creating and wo
   def product2(ns: List[Double]) =
     foldRight(ns, 1.0)(_ * _) // `_ * _` is more concise notation for `(x,y) => x * y`; see sidebar
 
+  def tail[A](l: List[A]): List[A] = l match {
+    //case Nil => sys.error("tail of Nil list")
+    case Nil => Nil
+    case Cons(x, xs) => xs
+  }
+
+  def setHead[A](l: List[A], h: A): List[A] = l match {
+    case Nil => Nil
+    case Cons(_, xs) => Cons(h, xs) 
+  }
+
+  def drop[A](l: List[A], n: Int): List[A] = {
+    if (n<=0) l
+    else l match {
+      case Nil => Nil
+      case Cons(x,xs) => drop(xs,n-1)
+    }
+  }
+
+  // def dropWhile[A](l: List[A], f: A => Boolean): List[A] = l match {
+  //   case Nil => Nil
+  //   case Cons(x,xs) =>
+  //     if (f(x))
+  //       dropWhile(xs, f)
+  //     else
+  //       l
+  // }
+
+  def dropWhile[A](l: List[A])(f: A => Boolean): List[A] = l match {
+    case Cons(h, t) if f(h) => dropWhile(t)(f)
+    case _ => l
+  }
+
+  def init[A](l: List[A]): List[A] = l match {
+    case Nil => Nil
+    case Cons(_,Nil) => Nil
+    case Cons(x, xs) => Cons(x, init(xs))
+  }
+
+  def length[A](l: List[A]): Int = {
+    foldRight(l, 0)((x,y) => y + 1)
+  }
+
+  def foldLeft[A,B](l: List[A], z: B)(f: (B, A) => B): B = {
+    @annotation.tailrec
+    def go[A,B](l: List[A], z: B, acc: B)(f: (B, A) => B): B = l match {
```

Comment:
The `z: B` parameter to the `go` function is redundant because it's not being processed in any way (it's not being used in any of the function logic).